### ISSUE TITLE
Nextcloud cannot be downgraded

### ISSUE THREAD

**Issue by Ekleog** [NixOS/nixpkgs#59025]
@Ma27, in https://github.com/NixOS/nixpkgs/issues/49783#issuecomment-480309261, mentions:
> (beware, downgrades aren't possible)

Is the upgrade properly gated behind `stateVersion`, then? I can't find any mention of `stateVersion` in the module, and it looks like there is a single `nextcloud` package.

And `nextcloud` has been bumped from 14 to 15 at least since 18.09.

I think we really need to backport to release-19.03 that `stateVersion` addition before 19.03 is released, by re-including nextcloud 14 and adding a `services.nextcloud.package` option that defaults to either nextcloud 14 or later depending on `stateVersion`.

@Ma27 Can you confirm or infirm this analysis?

**Comment by Ekleog** [comment]
Closing as `stateVersion` is apparently only for when manual intervention is required for `nixos-rebuild --upgrade`, not when it is required for `nixos-rebuild --rollback`.

While I'm sad about that, now is not a good time to change this.

**Comment by Ma27** [comment]
Not sure if it's still somehow relevant, but leaving some thoughts here:

> Is the upgrade properly gated behind stateVersion, then? I can't find any mention of stateVersion in the module, and it looks like there is a single nextcloud package.

How'd look such an approach in your opinion? As you've already mentioned we have a single `nextcloud` package that should be sufficiently tested by the maintainers before bumping and if you follow NixOS versions, you shouldn't have any problems.

The problem with keeping older versions is that they don't have a long [maintenance schedule](https://docs.nextcloud.com/server/15/admin_manual/release_schedule.html) and we'd end up having EOLed package versions in a stable NixOS version which is especially bad for software that is usually directly accessible from the internet and only protected by a simple login form.

I guess that this is mostly related to their database migrations, however most migration tools should support some kind of rollback functionality. But as most of their issues we faced here should be fixed upstream, unfortunately I lack energy and time to dig into their codebase and fix their stuff.

> Closing as stateVersion is apparently only for when manual intervention is required for nixos-rebuild --upgrade, not when it is required for nixos-rebuild --rollback.

I'm not sure if I understand this correctly, can you elaborate a bit more?

**Comment by Ekleog** [comment]
> How'd look such an approach in your opinion?

I think it might be based on `stateVersion` (or a per-module `stateEpoch` or similar as discussed in https://github.com/NixOS/nixpkgs/pull/50112), that would need to be bumped on each non-backwards-compatible change (in addition to each non-forwards-compatible change as it is now)

> As you've already mentioned we have a single nextcloud package that should be sufficiently tested by the maintainers before bumping

It is indeed, and we would need to keep multiple nextcloud packages if we want to do what I am thinking of.

> and if you follow NixOS versions, you shouldn't have any problems.

However, that is unfortunately not correct: when bumping a channel, I need to bump *all* modules. The example I personally bumped into was `opensmtpd` being bumped to >6.3, which changed the format of the configuration file. In order to handle this, it was required for me to adapt my configuration. After a few hours struggling with it, I gave up, and rollbacked to 18.09 for the time being. Had nextcloud been installed on the server on which I did this, the rollback would not have been possible (well, nextcloud would have broken)

It would in this specific case have been possible to use an overlay to recover the old opensmtpd without rolling the whole release back, but due to how nixos is architectured if such a change happened in a separate module it might very quickly become extremely complex to do that.

> The problem with keeping older versions is that they don't have a long maintenance schedule and we'd end up having EOLed package versions in a stable NixOS version which is especially bad for software that is usually directly accessible from the internet and only protected by a simple login form.

I think this should be handled as we handle unmaintained software with no major update: mark `meta.broken = true` even in the release branch, this way the users would be warned that their release is no longer maintained and would be given a choice of when to upgrade.

> I guess that this is mostly related to their database migrations, however most migration tools should support some kind of rollback functionality.

This would be the ideal solution indeed, but I don't think it's reasonable for us to fix all of our upstreams' issues with rolling back. The `stateVersion` (or similar) scheme would be used to say “within the same `stateVersion` it is always possible to upgrade and rollback, but when it is changed all bets are off, read the docs”.

>> Closing as stateVersion is apparently only for when manual intervention is required for nixos-rebuild --upgrade, not when it is required for nixos-rebuild --rollback.
>
> I'm not sure if I understand this correctly, can you elaborate a bit more?

Following a discussion on #nixos-dev on IRC, I've been pointed to the fact that currently `stateVersion` only means “within the same `stateVersion` it is always possible to upgrade”, and not “it is also possible to rollback” as I'd want it to.

I guess an RFC would be required to change that fact, hence my closing this issue :)
